Record Serialization

jhugard opened the pull request:
Work in Progress

Implement F# Record Serialization, per issue #38.
Also fixes #37, O(n^2) performance issue on repeated fields.

jhugard commented on Core/Hydration.fs:
```diff
@@ -0,0 +1,270 @@
+/// individual properties.  These can be used directly to create methods
+/// on classes, records, discriminated unions, etc.
+///
+/// In addition, a set of generic functions are provided which defines a
```
Delete this sentence.  These functions were moved to a separate file.

jhugard commented:
Retrigger build: Travis-CI seems flaky.

jhugard commented:
@ctaggart I am seeing occasional and inconsistent build failures running on x86-64 ArchLinux with 2 cores and the latest mono (4.4.0).  It looks like a race condition in mono itself or in xunit.

Disabling parallel execution of the tests (`Parallel = NoParallelization` in build.fsx) caused different issues: first run caused a hang, and second run got a different error (second error listed below).

Can you look over the code and see if you spot any obvious badness?  Could it be initialization of static members on the serialization classes/records?

*Failure in the build, reproduced locally; same as travis-ci build failure*
```
xUnit.net Console Runner (64-bit .NET 4.0.30319.42000)
  Discovering: Froto.Parser.Test
  Discovering: Froto.Core.Test
  Discovered:  Froto.Core.Test
  Starting:    Froto.Core.Test
  Discovered:  Froto.Parser.Test
  Starting:    Froto.Parser.Test
  Finished:    Froto.Core.Test
  Finished:    Froto.Parser.Test
System.TypeInitializationException: The type initializer for 'System.Collections.Generic.List`1' threw an exception. ---> System.Threading.ThreadAbortException:
  --- End of inner exception stack trace ---
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x40d0a4f0 + 0x00098> in <filename unknown>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x40d03b50 + 0x001fd> in <filename unknown>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x40d03b10 + 0x00023> in <filename unknown>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x40d0a440 + 0x0005b> in <filename unknown>:0
  at System.Threading.ThreadHelper.ThreadStart () <0x40df1020 + 0x00033> in <filename unknown>:0
*** Error in `/usr/bin/mono': double free or corruption (out): 0x00007fed4800cbe0 ***

```

*Failure in the build with XUnit parallelization disabled*
```
xUnit.net Console Runner (64-bit .NET 4.0.30319.42000)
  Discovering: Froto.Parser.Test
  Discovered:  Froto.Parser.Test
  Starting:    Froto.Parser.Test
  Finished:    Froto.Parser.Test
  Discovering: Froto.Core.Test
Stacktrace:

  at <unknown> <0xffffffff>
  at System.Collections.Concurrent.ConcurrentDictionary`2<TKey_REF, TValue_REF>.GrowTable (System.Collections.Concurrent.ConcurrentDictionary`2/Tables<TKey_REF, TValue_REF>,System.Collections.Generic.IEqualityComparer`1<TKey_REF>,bool,int) <0x00614>
  at System.Collections.Concurrent.ConcurrentDictionary`2<TKey_REF, TValue_REF>.TryAddInternal (TKey_REF,TValue_REF,bool,bool,TValue_REF&) <0x0060b>
  at System.Collections.Concurrent.ConcurrentDictionary`2<TKey_REF, TValue_REF>.GetOrAdd (TKey_REF,System.Func`2<TKey_REF, TValue_REF>) <0x00087>
  at Xunit.Sdk.CollectionPerClassTestCollectionFactory.Get (Xunit.Abstractions.ITypeInfo) <0x00163>
  at Xunit.Sdk.XunitTestFrameworkDiscoverer.CreateTestClass (Xunit.Abstractions.ITypeInfo) <0x00029>
  at Xunit.Sdk.TestFrameworkDiscoverer/<>c__DisplayClass26_0.<Find>b__0 () <0x00168>
  at Xunit.Sdk.XunitWorkerThread/<>c.<QueueUserWorkItem>b__4_0 (object) <0x00052>
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (object) <0x00049>
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x001f9>
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x00023>
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x0004b>
  at System.Threading.ThreadPoolWorkQueue.Dispatch () <0x001e4>
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x0000b>
  at (wrapper runtime-invoke) <Module>.runtime_invoke_bool (object,intptr,intptr,intptr) <0x0005a>
```

jhugard commented:
@ctaggart Accidently renamed Froto.Core.DLL to Froto.Encoding.DLL in the latest commit (yay global-search-and-replace!)  What do you think about keeping it that way?  Change "Froto.Core" project to "Froto.Encoding" and put all the encoding and serialization functions in that namespace?

ctaggart commented on Froto.sln:
```diff
@@ -22,6 +22,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		build.fsx = build.fsx
 		paket.dependencies = paket.dependencies
 		paket.lock = paket.lock
+		Performance1.psess = Performance1.psess
```
This snuck back in.

jhugard replied:
I'll remove this shortly.

ctaggart commented:
@jhugard Scrolled through and it looks good. Were you wanting to change the assembly names too or leave it as `Froto.Core`?

jhugard commented:
@ctaggart Cool, thanks for having a look!  I'll remove the performance config shortly.

Do you have an opinion on renaming the assembly?  I'm fine either way, but "Core" isn't very descriptive; it's not actually needed for parsing and manipulating proto files (without doing any serialization).

ctaggart commented:
I think `Froto.Serialization` makes sense. It matches the namespace you've given it as well as the documentation.
https://developers.google.com/protocol-buffers/docs/overview#what-are-protocol-buffers

`Serialization` is what others have used:
https://www.nuget.org/packages/Finsa.CodeServices.Serialization.Protobuf/
https://www.nuget.org/packages/XLabs.Serialization.ProtoBuf/

jhugard commented:
Support for Record Serialization complete (!!!)
Also cleaned up and reorganized the namespaces/modules, renamed Froto.Core.dll -> Froto.Serialization.dll, froto.exe -> frotoc.exe, and renamed the project directories for consistency.

@ctaggart Ready for final review before merging.

jhugard commented:
Summary of changes:
* Implemented support for serializing F# Records
* Reimplemented Class serialization in terms of the new encode/decode structure developed for Record Serialization
* Significant reorganization of serialization namespaces, modules, and function names
   * At the WireFormat level, renamed encodeXXX/decodeXXX functions to packXXX/unpackXXX
   * At the RawField level, and renamed dehydrateXXX/hydrateXXX functions to Encode.fromXXX and Decode.toXXX
   * At the Serialization level, created functions for serializing to/from ZeroCopyBuffer and ArraySegment
   * Renamed base namespace from Froto.Core to Froto.Serialization
   * Renamed assembly from Froto.Core.dll to Froto.Serialization.dll
   * Renamed “Exe” solution to “Froto.Compiler.Exe” and renamed “froto.exe” to “frotoc.exe”
   * Renamed several folders for naming consistency; e.g., “ProtoParser” to “Parser” and “Exe” to “Compiler.Exe”

jhugard commented:
Open questions before merging this to `master`:

* Should the pack/unpack functions follow the same pattern as the Encode/Decode functions?  Or, is it  good enough?
   *  E.g., rather than unpackVarint, use Unpack.toVarint?
* Should the Serialization functions follow the same Module pattern?  Or, is it good enough?
   * I.e., create a module to house the ZeroCopyBuffer serialization functions, and another for the ArraySegment functions?  Right now, they are all in the Serialization module.  I don’t have any good ideas on what to call the modules, though: something like “Serializer.ZeroCopyBuffer.serializeTo” feels… a bit unwieldy.
* Should the release notes include more detail?  Should this be done before merge?  Or, just defer until docs are written?

ctaggart commented:
Looks like you went with the modules. I'm going to trust your judgement on these questions.

jhugard commented:
@ctaggart 
* On the exe name, just trying to follow industry practice for naming compilers (_lang_*C*.exe).
* On the project name, was also trying to follow industry practice, which is to put the heavy lifting in one project (e.g., Froto.Compiler) and put the EXE shell into another project (e.g., Froto.Compiler.Exe).  However, that might just be a C++ thing, as I haven't examined a lot of recent .NET projects.

Please confirm you'd prefer `froto.exe` and prefer it living in `Froto.Compiler`; I can make those changes tonight...

ctaggart commented:
I understand that logic for a compiler, but I've always found it strange that they consider `protoc` a compiler. Does it produce binaries? `froto.exe` was intended to just do code generation.

jhugard commented:
@ctaggart Ok, renamed things as requested (I think):

* Namespace: Froto.Compiler.Exe -> Froto.Compiler
* Directory: Compiler.Exe -> Compiler
* Binary: frotoc.exe -> froto.exe

But, that's somewhat inconsistent because the directory and project are still Compiler and Froto.Compiler, respectively.

IMHO, the term Compiler is technically correct here (see [Wikipedia](https://en.wikipedia.org/wiki/Compiler)): the Froto program will be doing lexing, parsing, semantic analysis, and code generation.  However, let me know if you'd prefer to change it to one of the following more specific terms, or let the last commit stand and continue with the merge.

Project:
- [ ] Froto.Compiler (leave as-is - just do the merge already! lol)
- [ ] Froto.Transpiler
- [ ] Froto.CodeGen
- [ ] Froto.Exe (put it back to what it was)
- [ ] Froto.Compiler.Exe

Binary:
- [ ] froto.exe (leave as-is - just do the merge already!)
- [ ] frotogen.exe
- [ ] frotoc.exe

ctaggart commented:
Makes sense. I think you can leave it as is and merge it in both cases.

ctaggart commented on release_notes.md:
```diff
@@ -2,6 +2,16 @@
 ### 0.3.0 _ under construction
   * [#3](https://github.com/ctaggart/froto/issues/3) F# type provider
   * [#15](https://github.com/ctaggart/froto/issues/15) froto.exe code generator for proto3
+  * [#37](https://github.com/ctaggart/froto/issues/38) Fix bug: O(n^2) Performance Problem on repeated fields
```
#3 and #15 aren't actually done. We can make a 0.3.0 release and bump these to 0.4.0.

jhugard commented:
Already did the merge.  Do you mind fixing up the release notes and doing the release-thing?

ctaggart commented:
Yeah, I'll do it tomorrow morning.

jhugard commented:
Ok... I might have another PR merged in by then - thought I was done, but a refinement just hit me that I like for organizing the `Serializer` functions.

Rather than `Serializer.serializeTo` for ZeroCopyBuffer and `Serializer.serialize` for ArraySegment, I want to try `Serialize.toZeroCopyBuffer` and `Serialize.toArraySegment`, etc.

In the future, adding for example `Serialize.toArray` and `Serialize.toStream` would be clean and clear.

jhugard commented:
See PR #40, which has already been merged to master.  I think it's a solid improvement.
I've also created a release and [v0.3.0 release tag](https://github.com/ctaggart/froto/releases).
Not sure what else you need to publish to NuGet, but everything else should be all ready!

ctaggart commented:
Yeah, pretty sure the tags are not supposed to have the `v`. I'll check.
![image](https://cloud.githubusercontent.com/assets/80104/16089870/f372a64c-32e1-11e6-8a88-22933f738233.png)

ctaggart commented:
Yes, that was it. Fixed & uploaded. Great work!
https://ci.appveyor.com/project/ctaggart/froto/build/0.3.0
https://www.nuget.org/packages/Froto.Parser
https://www.nuget.org/packages/Froto.Serialization

jhugard commented:
Not sure if you saw it and not sure if you monitor your gmail acct, but I did some work on Parsing today and would like to get release 0.3.1 out asap. Already merged and tagged...

Sent from my iPhone

On Jun 15, 2016, at 10:33 AM, Cameron Taggart <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, that was it. Fixed & uploaded. Great work!
https://ci.appveyor.com/project/ctaggart/froto/build/0.3.0
https://www.nuget.org/packages/Froto.Parser
https://www.nuget.org/packages/Froto.Serialization

-
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/ctaggart/froto/pull/39#issuecomment-226261192>, or mute the thread<https://github.com/notifications/unsubscribe/AAPT0kbw0kAFwJ1V0eyHFpmbXvKGFNLyks5qMDd-gaJpZM4IlqWl>.

ctaggart commented:
Currently driving for next 5 hours. I'll add you as a publisher when I
publish these.
On Jun 15, 2016 4:45 PM, "James Hugard" <notifications@github.com> wrote:

> Not sure if you saw it and not sure if you monitor your gmail acct, but I
> did some work on Parsing today and would like to get release 0.3.1 out
> asap. Already merged and tagged...
>
> Sent from my iPhone
>
> On Jun 15, 2016, at 10:33 AM, Cameron Taggart <notifications@github.com
> <mailto:notifications@github.com>> wrote:
>
>
> Yes, that was it. Fixed & uploaded. Great work!
> https://ci.appveyor.com/project/ctaggart/froto/build/0.3.0
> https://www.nuget.org/packages/Froto.Parser
> https://www.nuget.org/packages/Froto.Serialization
>
> -
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub<
> https://github.com/ctaggart/froto/pull/39#issuecomment-226261192>, or
> mute the thread<
> https://github.com/notifications/unsubscribe/AAPT0kbw0kAFwJ1V0eyHFpmbXvKGFNLyks5qMDd-gaJpZM4IlqWl
> >.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ctaggart/froto/pull/39#issuecomment-226351995>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAE46Awp1wVLoaEjJG0K2rhoCDgyf3-Eks5qMI6hgaJpZM4IlqWl>
> .
>

jhugard commented:
Excellent!

Have a safe drive. Stop texting.

:)

Sent from my iPhone

On Jun 15, 2016, at 4:52 PM, Cameron Taggart <notifications@github.com<mailto:notifications@github.com>> wrote:

Currently driving for next 5 hours. I'll add you as a publisher when I
publish these.
On Jun 15, 2016 4:45 PM, "James Hugard" <notifications@github.com<mailto:notifications@github.com>> wrote:

> Not sure if you saw it and not sure if you monitor your gmail acct, but I
> did some work on Parsing today and would like to get release 0.3.1 out
> asap. Already merged and tagged...
>
> Sent from my iPhone
>
> On Jun 15, 2016, at 10:33 AM, Cameron Taggart <notifications@github.com<mailto:notifications@github.com>
> <mailto:notifications@github.com>> wrote:
>
>
> Yes, that was it. Fixed & uploaded. Great work!
> https://ci.appveyor.com/project/ctaggart/froto/build/0.3.0
> https://www.nuget.org/packages/Froto.Parser
> https://www.nuget.org/packages/Froto.Serialization
>
> -
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub<
> https://github.com/ctaggart/froto/pull/39#issuecomment-226261192>, or
> mute the thread<
> https://github.com/notifications/unsubscribe/AAPT0kbw0kAFwJ1V0eyHFpmbXvKGFNLyks5qMDd-gaJpZM4IlqWl
> >.
>
> -
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ctaggart/froto/pull/39#issuecomment-226351995>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAE46Awp1wVLoaEjJG0K2rhoCDgyf3-Eks5qMI6hgaJpZM4IlqWl>
> .
>

-
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/ctaggart/froto/pull/39#issuecomment-226352982>, or mute the thread<https://github.com/notifications/unsubscribe/AAPT0re6PLX6QNw06pHP9oeT2OsQV2U2ks5qMJAfgaJpZM4IlqWl>.
